expression/builtin_control.go

```diff
@@ -125,32 +130,102 @@ type ifNullFunctionClass struct {
 	baseFunctionClass
 }
 
-func (c *ifNullFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinIfNullSig{newBaseBuiltinFunc(args, ctx)}
-	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
+func (c *ifNullFunctionClass) getFunction(args []Expression, ctx context.Context) (sig builtinFunc, err error) {
+	if err := errors.Trace(c.verifyArgs(args)); err != nil {
+		return nil, errors.Trace(err)
+	}
+	tp0, tp1 := args[0].GetType(), args[1].GetType()
+	fieldTp := types.AggFieldType([]*types.FieldType{tp0, tp1})
+	types.SetBinChsClnFlag(fieldTp)
+	classType := types.AggTypeClass([]*types.FieldType{tp0, tp1}, &fieldTp.Flag)
+	fieldTp.Decimal = mathutil.Max(tp0.Decimal, tp1.Decimal)
+	// TODO: make it more accurate when inferring FLEN
+	fieldTp.Flen = tp0.Flen + tp1.Flen
+
+	var evalTps evalTp
+	switch classType {
+	case types.ClassInt:
+		evalTps = tpInt
+		fieldTp.Decimal = 0
+	case types.ClassReal:
+		evalTps = tpReal
+	case types.ClassDecimal:
+		evalTps = tpDecimal
+	case types.ClassString:
+		evalTps = tpString
+		if !types.IsBinaryStr(tp0) && !types.IsBinaryStr(tp1) {
+			fieldTp.Charset, fieldTp.Collate = mysql.DefaultCharset, mysql.DefaultCollationName
+			fieldTp.Flag ^= mysql.BinaryFlag
+		}
+	}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, evalTps, evalTps, evalTps)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	bf.tp = fieldTp
+	switch classType {
+	case types.ClassInt:
+		sig = &builtinIfNullIntSig{baseIntBuiltinFunc{bf}}
+	case types.ClassReal:
+		sig = &builtinIfNullRealSig{baseRealBuiltinFunc{bf}}
+	case types.ClassDecimal:
+		sig = &builtinIfNullDecimalSig{baseDecimalBuiltinFunc{bf}}
+	case types.ClassString:
+		sig = &builtinIfNullStringSig{baseStringBuiltinFunc{bf}}
+	}
+	return sig.setSelf(sig), nil
 }
 
-type builtinIfNullSig struct {
-	baseBuiltinFunc
+type builtinIfNullIntSig struct {
+	baseIntBuiltinFunc
 }
 
-// eval evals a builtinIfNullSig.
-// See https://dev.mysql.com/doc/refman/5.7/en/control-flow-functions.html#function_ifnull
-func (b *builtinIfNullSig) eval(row []types.Datum) (types.Datum, error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return types.Datum{}, errors.Trace(err)
+func (b *builtinIfNullIntSig) evalInt(row []types.Datum) (int64, bool, error) {
+	sc := b.ctx.GetSessionVars().StmtCtx
+	arg0, isNull, err := b.args[0].EvalInt(row, sc)
+	if !isNull {
```

Comment:
we do not check err here,
we return errors.Trace(err) in next line,
if err == nil , errors.Trace(err) will be null.